### DISCUSSION TITLE
Deployment process for local code

### DISCUSSION THREAD

**Discussion by Bouchemel-Nasreddine** [ontoportal/ontoportal-project#19] [Q&A] [answered]
Hey everybody, i'm Nasr, new developer in the IndustryPortal team, i was looking through the Ontoportal documentation, exactly in the "Advanced Configuration" page, I've been trying to deploy new version of the project, both on the VM, and our test server, for the VM, i think i should change a bit in the configs files to link the projects there to the IndustryPortal ones, so i wanted to know how exactly i could do that? thnx in advance

**Comment by syphax-bouazzouni** [comment, marked as answer]
Would you like to deploy the UI or the API ? 

And ping @alexskr that may help with this.

**Reply by Bouchemel-Nasreddine** [reply]
for now i'm working on the UI side, @alexskr could you help me with this?

**Reply by syphax-bouazzouni** [reply]
Is your code pushed on GitHub?

**Reply by Bouchemel-Nasreddine** [reply]
yes it is pushed, then i went to the setup_deploy_env.sh file, i changed the repo url "github.com/ncbo/bio..." with the industryportal one, but it didn't affect anything

**Reply by syphax-bouazzouni** [reply]
I can give you a temporary and hacky solution that did work with your predecessor (Abdel), do the following (test it before in your VM)

```bash 
cd /srv/ontoportal/bioportal_web_ui/current
git init
git remote add origin <URL of your UI repository>
git pull
sudo oprestart
```
Tell if it worked, after the `git pull` you will surelly update the config files (`config/bioportal_config_appliance.rb`)

We will work in the future, to make the deployment easier and better

**Reply by Bouchemel-Nasreddine** [reply]
yes i'm working in the VM, but first should i do it as root or ontoportal user?

**Reply by syphax-bouazzouni** [reply]
ontoportal user

**Reply by Bouchemel-Nasreddine** [reply]
i pulled the repo and did oprestart, but it gaves me a error "Web application could not be started by the Phusion Passenger(R) application server", 
NB: running the deployment scripts (deploy_ui.sh) gets the app back to its original state
![Screenshot from 2023-02-14 12-03-55](https://user-images.githubusercontent.com/45484784/218719619-db37e34f-93fa-4afd-862c-d2a48ed95270.png)

**Reply by syphax-bouazzouni** [reply]
you can see logs in `/var/log/httpd` to know more about what is the issue in your code or configuration. (use `ls -ltr` to find the latest updated log)

**Reply by Bouchemel-Nasreddine** [reply]
hey again @syphax-bouazzouni , I've been looking at logs and I've come up with a solution and the site shows up now, though the problem with it, despite having all the UI project files updated, the site is messing some layouts and some configs, here is a picture so you can understand more: 
![Screenshot from 2023-02-15 11-09-31](https://user-images.githubusercontent.com/45484784/218998298-8705496d-855a-4c9b-b125-f93eadd4f895.png)
 do you have any ideas about this?

**Reply by syphax-bouazzouni** [reply]
So what was the error in the log? and how did you solve it?

For Messing layouts and configs it is hard to tell without more information, but I guess that it doesn't use your configuration file `config/bioportal_config_development.rb`. Because in the appliance, the environment is `appliance` not 'development` like in your local machine. 

This means you need to update also the file `config/bioportal_config_appliance.rb` in your appliance (and server) to make it use it. (and restart it)

**Reply by Bouchemel-Nasreddine** [reply]
well, at first it needed a `bundle install`, then i needed to add `config.assets.compile = true` to the `config/environments/appliance.rb` and that's it.
now i modified the appliance config files, it has the team members and some other stuff, the thing is even when i set the `UI_THEME` variable in the configs to `:industruportal`, it won't change

**Reply by syphax-bouazzouni** [reply]
if you try `sudo opclearcaches`  in your appliance/server does it work?

**Reply by Bouchemel-Nasreddine** [reply]
tried it then `sudo oprestart`, but it remains the same

**Reply by syphax-bouazzouni** [reply]
Okay, we can schedule a fast meeting if you want; tell me in Slack when you are available.

**Reply by Bouchemel-Nasreddine** [reply]
hello again, after successfully deploying the project on a VM, i tried to do the same on the test server, but it didn't work and i think it's a problem of version of bunlder, with i tap `bundle -v` in the ontoportal user i get version 2.1.4, despite having the version 2.3.15 in the centos user, any ideas?

Edit:
it's done the ontoportal user didn't have some permissions, so i excuted the installation command with a flag `bundle install bundler --user-install`

**Reply by Bouchemel-Nasreddine** [reply]
after executing `RAILS_ENV=appliance RAILS_SERVE_STATIC_FILES=1 RAILS_LOG_TO_STDOUT=1 bin/rake assets:precompile`, i get an error saying 
`NoMethodError: undefined method represent_boolean_as_integer=' for nil:NilClass
/srv/ontoportal/bioportal_web_ui/releases/20230216204533/config/initializers/new_framework_defaults_5_2.rb:32:in <top (required)>'`
i have ruby version 2.7.0 and rails 7.0.3

**Reply by syphax-bouazzouni** [reply]
the `new_framework_defaults_5_2` does not exist in your code https://github.com/IndustryPortal/bioportal_web_ui/tree/master/config/initializers 

are you sure you have the correct code in your appliance?

**Reply by Bouchemel-Nasreddine** [reply]
yes
![ss](https://user-images.githubusercontent.com/45484784/219597365-96f67637-d593-4331-a518-512bd2f47794.png)

**Reply by syphax-bouazzouni** [reply]
I don't think you have the correct code in your server (or maybe it got corrupted with the `deploy_ui.sh` script).
Re-run the following
```bash
git stash #will clear your changes
git pull #will get your latest version of the code
RAILS_ENV=appliance RAILS_SERVE_STATIC_FILES=1 RAILS_LOG_TO_STDOUT=1 bin/rake assets:precompile # will generate your assets 
sudo oprestart 

```

**Reply by Bouchemel-Nasreddine** [reply]
i went to the initializes folder and deleted the 5_2 file, i don't really know why it was there in the first place, despite not having it in git repo, the command went successfully, and now i have this error in the httpd/error_log file saying  `Error: The application encountered the following error: https://github.com/ontoportal-lirmm/ontologies_api_ruby_client.git (at master@fe929a4) is not yet checked out. Run `bundle install` first`, i executed bundle install as the error says but with no result

**Reply by syphax-bouazzouni** [reply]
So the issue was in your configured ruby in the version in `/etc/httpd/conf.d/10-appliance.ontoportal.org_non-tls.conf`, putting the correct  version made it work

**Reply by Bouchemel-Nasreddine** [reply]
yes exactly, thank you Syphax

**Comment by RaimiSol** [comment]
Hello I'm the developer from @ontoportal/biodivportal. I'm currently trying to deploy the ontologies_api with the agroportal branch. 

@Bouchemel-Nasreddine do you managed to deploy ontologies_api? In the case of deployment: what was your approach?
I'm documenting my deployment steps here if you're interested: https://docs.google.com/document/d/14vKMva_5y8XtPXiV6ChN-pSScU0eo0waXWl-tmgv-e4/edit?usp=sharing

**Reply by Bouchemel-Nasreddine** [reply]
in my case i deployed the UI, with the approch that @syphax-bouazzouni has mentionned, i can see in your doc file the these are the steps that were needed in my case

**Reply by RaimiSol** [reply]
Thanks for the reply. 

If I understand correctly, you don't have deployed the backend (ontologies_api) yet?

**Reply by Bouchemel-Nasreddine** [reply]
yes, till now i didn't need any new version of API, the previously deployed one on our servers,  fits my requirements

**Reply by RaimiSol** [reply]
Ok I have managed to deploy the backend (ontologies_api). For any questions regarding my approach contact me or refer to this document: https://docs.google.com/document/d/14vKMva_5y8XtPXiV6ChN-pSScU0eo0waXWl-tmgv-e4/edit?usp=sharing
